sim/warlock/talents.go

```diff
@@ -578,10 +578,19 @@ func (warlock *Warlock) setupDemonicPact() {
 			newSPBonus := 0.0
 			if warlock.DemonicPactAura.IsActive() {
 				lastBonus = warlock.DemonicPactAura.ExclusiveEffects[0].Priority
-				newSPBonus = math.Floor(lastBonus*dpMult*dpMult+
-					(1-dpMult)*dpMult*warlock.GetStat(stats.SpellPower)) + 1
+
+				if warlock.Options.NewDPBehaviour {
+					newSPBonus = math.Round(dpMult * (warlock.GetStat(stats.SpellPower) - lastBonus))
+				} else {
+					newSPBonus = math.Floor(lastBonus*dpMult*dpMult+
+						(1-dpMult)*dpMult*warlock.GetStat(stats.SpellPower)) + 1
+				}
 			} else {
-				newSPBonus = math.Floor(warlock.GetStat(stats.SpellPower)*(dpMult+dpMult*dpMult)) + 1
+				if warlock.Options.NewDPBehaviour {
+					newSPBonus = math.Round(dpMult * warlock.GetStat(stats.SpellPower))
+				} else {
+					newSPBonus = math.Floor(warlock.GetStat(stats.SpellPower)*(dpMult+dpMult*dpMult)) + 1
+				}
```

Comment:
I think this would be simpler and easier to reason about correctly if we branch on the new option first:

```
    lastBonus := 0.0
    if warlock.DemonicPactAura.IsActive() {
        lastBonus = warlock.DemonicPactAura.ExclusiveEffects[0].Priority
    }
  
    newSPBonus := 0.0

    if warlock.Options.NewDPBehaviour {
        newSPBonus = math.Round(dpMult * warlock.GetStat(stats.SpellPower))
    } else {
        if warlock.DemonicPactAura.IsActive() {
            newSPBonus = math.Floor(lastBonus*dpMult*dpMult+
						(1-dpMult)*dpMult*warlock.GetStat(stats.SpellPower)) + 1
        } else {
             newSPBonus = math.Floor(warlock.GetStat(stats.SpellPower)*(dpMult+dpMult*dpMult)) + 1
        }
    }


}